build-system/tasks/helpers.js

```diff
@@ -479,13 +486,18 @@ async function compileUnminifiedJs(srcDir, srcFilename, destDir, options) {
       };
 
       build.onLoad({filter: /.*/, namespace: ''}, async (file) => {
-        const contents = await fs.promises.readFile(file.path, 'utf-8');
-        const hash = crypto.createHash('sha1').update(contents).digest('hex');
-        if (cache.has(hash)) {
-          return {contents: cache.get(hash)};
+        const relativePath = path.relative(
+          path.resolve(__dirname, '..', '..'),
+          file.path
+        );
+        const cachedContents = cache.get(relativePath);
+        if (cachedContents) {
+          return {contents: cachedContents};
         }
+
+        const contents = await fs.promises.readFile(file.path, 'utf-8');
         const transformed = await transformContents({file, contents});
-        cache.set(hash, transformed.contents);
+        cache.set(relativePath, transformed.contents);
```

Comment:
This logic has changed with #32891. We now cache unresolved promises instead of the transformation results in order to account for how `esbuild` does a whole bunch of `onLoad` calls in parallel. Will need to rethink how to invalidate the cache, but I'm glad you've raised the point.